src/encoded/types/cart.py

```diff
@@ -48,51 +54,94 @@ class Collection(Item.Collection):
-            'elements': []
-        }
+        initial_cart,
     )
-    return request.resource_path(cart)
+    cart_path = request.resource_path(cart)
+    return cart_path
 
 
-@view_config(context=Cart.Collection, request_method='GET', permission='save-carts', name='get-cart')
-def get_or_create_cart_by_user(context, request):
+def _get_userid(request):
     userid = [
         p.replace('userid.', '')
         for p in request.effective_principals
         if p.startswith('userid.')
     ]
     if not userid:
         raise HTTPBadRequest()
-    else:
-        userid = userid[0]
+    return userid[0]
+
+
+def _get_user(request, userid):
     user = request.registry[COLLECTIONS]['user'].get(userid)
     if not user:
         raise HTTPBadRequest()
-    carts = _get_carts_by_user(request, userid)
+    return user
+
+
+@view_config(context=Cart.Collection, request_method='GET', permission='save-carts', name='get-cart')
+def get_or_create_cart_by_user(context, request):
+    userid = _get_userid(request)
+    user = _get_user(request, userid)
+    carts = get_cart_objects_by_user(request, userid, ['disabled', 'deleted '])
     if not carts:
         cart = _create_cart(request, user)
+        cart_atids = None
+    else:
+        cart_atids = [cart['@id'] for cart in carts]
+    request.response.status = 200
+    return {
+        'status': 'success',
+        '@type': ['result'],
+        '@graph': cart_atids or [cart]
+    }
+
+
+@view_config(context=Cart.Collection, request_method='PUT', permission='save-carts', name='put-cart')
+def create_cart_by_user(context, request):
+    userid = _get_userid(request)
+    user = _get_user(request, userid)
+    user_carts = get_cart_objects_by_user(request, userid, ['deleted'] if not 'group.admin' in request.effective_principals else [])
+    cart_atids = [cart['@id'] for cart in user_carts]
+    cart_status = request.json.get('status', None)
+    cart_name = request.json.get('name', '').strip()
+    # User writing a new cart; check for cart overflow and naming conflicts
+    if cart_status != 'disabled':
+        countable_carts = [
+            cart
+            for cart in user_carts
+            if cart['status'] != 'disabled' and cart['status'] != 'deleted'
+        ]
+    else:
+        # Creating an auto-save cart, so ignore existing carts
+        countable_carts = []
+    if len(countable_carts) >= CART_USER_MAX:
+        msg = 'Users cannot have more than {} carts'.format(CART_USER_MAX)
+        raise HTTPBadRequest(explanation=msg)
+    elif next((cart for cart in user_carts if (cart['status'] != 'deleted' and
+            cart['status'] != 'disabled' and
+            cart['name'].strip().upper() == cart_name.upper())), None):
+        msg = 'A cart with the name "{}" already exists'.format(cart_name)
+        raise HTTPBadRequest(explanation=msg)
+    cart_identifier = request.json.get('identifier', None)
+    cart = _create_cart(request, user, cart_name, cart_identifier, cart_status)
```

Comment:
@caseylitton also points out we shouldn't be passing keyword arguments as position arguments.

```python
cart = _create_cart(
   request,
   user,
   name=cart_name,
   identifier=cart_identifier,
   status=cart_status
)
```

Reply:
Changed